### PR TITLE
Allow reward functions to log extra columns and scalar metrics

### PR DESCRIPTION
**Pull request by manueldeprada**
## TLDR

Adds `log_extra` and `log_metric` hooks to reward functions in GRPO and RLOO so custom reward functions can log extra completion columns and scalar metrics without accessing private trainer state.

image
image

## So:

While working with custom reward functions (e.g. for calibration training), I found myself using `log_completions=True`, which logs samples to trackio (super useful!). 

However, I also needed to log additional columns to the completions table (like extracted answers and gold labels) and scalar metrics (like accuracy and ECE) directly from within the reward function. Currently there’s no clean way to do this in GRPO without accessing private state, to the best of my knowledge.

This PR adds two public methods to `_BaseTrainer` and passes them to reward functions via `reward_kwargs`:

* `log_extra(column, values)`: adds extra columns to the completions table (parquet + wandb/trackio)
* `log_metric(name, value)`: logs scalar metrics through the existing `_metrics` mechanism, so they show up as plots alongside KL, entropy, etc.

Both are fully backwards compatible, as existing reward functions will just absorb them into `**kwargs`. It is not super orthodox to pass functions like this, but it is simple and adds very little code.. I also added them to RLOO, as it is similar, for completeness.

## Example usage

```python
def my_reward_fn(completions, answer, log_extra=None, log_metric=None, **kwargs):
    extracted = [extract_answer(c) for c in completions]
    rewards = [1.0 if e == a else 0.0 for e, a in zip(extracted, answer)]

    if log_extra:
        log_extra("golden_answer", list(answer))
        log_extra("extracted_answer", extracted)

    if log_metric:
        log_metric("accuracy", sum(rewards) / len(rewards))

    return rewards
```

Docs, tests, and signatures were Clauded, so lmk if you’d be interested in merging it and I can put more effort into polishing the PR.

Shoutout to the berner guys, Andy, @lewtun, joel and @lvwerra. Hope everything is going well, its also fun to find myself using TRL from the other side 🙂

---

> [!NOTE]
> **Medium Risk**
> Touches GRPO/RLOO training-time logging and distributed gather/aggregation logic; mistakes could misattribute logged columns/metrics across ranks or impact logging output, but core optimization/reward computation is unchanged.
> 
> **Overview**
> **Custom reward functions can now emit richer logs during GRPO/RLOO training.** Both trainers pass two new kwargs into reward functions: `log_extra(column, values)` for adding per-completion columns to the completions table and `log_metric(name, value)` for emitting scalar metrics.
> 
> The trainers buffer these user logs, **gather/average them across distributed processes in a deterministic (sorted-key) order**, and include them in the parquet/trackio/wandb completions outputs and the standard `log_history` metric stream.
> 
> Docs for `GRPOTrainer`/`RLOOTrainer` add the new kwargs and an Example 6, and tests add coverage that `log_extra` populates `trainer._logs["extra"]` and `log_metric` appears in `state.log_history` (with minor test config tweaks to reduce memory).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 46e49aa351615a032ba87c47943bef1fcc59a420. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/trainer/rloo_trainer.py [hunk 1]
```diff
@@ -814,6 +815,12 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
         # This allows for dynamic reward shaping based on training progress.
         reward_kwargs["trainer_state"] = self.state
 
+        # Allow reward functions to log extra columns to the completions table.
+        reward_kwargs["log_extra"] = self.log_completion_extra
+
+        # Allow reward functions to log additional scalar metrics.
+        reward_kwargs["log_metric"] = self.log_metric
+
```
[the remark this answers is not in the document]
Comment (manueldeprada) [reply trl/trainer/rloo_trainer.py:827]: fixed

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1140,6 +1141,12 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
         # This allows for dynamic reward shaping based on training progress.
         reward_kwargs["trainer_state"] = self.state
 
+        # Allow reward functions to log extra columns to the completions table.
+        reward_kwargs["log_extra"] = self.log_completion_extra
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: fair, I'll fix it

**Comment by qgallouedec** [comment]
Thanks @manueldeprada!

I understand the need, there's currently no simple way to log extra metrics from within a reward function without modifying the codebase. The closest workaround today is to use separate reward functions with `reward_weights=0` for pure logging, but that doesn't help when the metric is a byproduct of the reward computation and you want to avoid recomputing it.

That said, I'm not fully sold on the approach. A reward function should ideally be passive: it takes inputs, returns a scalar, and has no side effects. Passing a `log_metric` callback blurs that boundary.

I don't have a better alternative in mind though, and from the user's perspective, this is probably the most intuitive and simple solution. So I'd consider merging it. Open to thoughts from others. @albertvillanova @AmineDiro.

**Comment by qgallouedec** [comment]
@codex review

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1142,6 +1171,12 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
         # This allows for dynamic reward shaping based on training progress.
         reward_kwargs["trainer_state"] = self.state
 
+        # Allow reward functions to log extra columns to the completions table.
+        reward_kwargs["log_extra"] = self._log_completion_extra
+
+        # Allow reward functions to log additional scalar metrics.
+        reward_kwargs["log_metric"] = self._log_metric
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:1165]: No, this is a discouraged practice (not to use kwargs), precisely for forward compatibility

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1939,6 +1976,23 @@ def _generate_and_score_completions(
...
+        self._pending_extra_logs.clear()
+
+        # Flush user-logged metrics (from log_metric), averaging across processes.
+        # Keys must be sorted so that all ranks call accelerator.gather in the same order, otherwise values
+        # get mis-attributed across metrics (dict insertion order may differ between processes).
+        for name in sorted(self._pending_metrics):
+            values = self._pending_metrics[name]
+            local_mean = sum(values) / len(values)
+            global_mean = self.accelerator.gather(torch.tensor(local_mean, device=device)).mean().item()
+            self._metrics[mode][name].append(global_mean)
```
[the remark this answers is not in the document]
Comment (manueldeprada) [reply trl/trainer/grpo_trainer.py:2014]: I would dismiss this. If a user calls log_metric("reward", ...), that's on them. This is a feature for advanced users anyway.

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -2319,6 +2373,7 @@ def log(self, logs: dict[str, float], start_time: float | None = None) -> None:
                 "prompt": self._logs["prompt"],
                 "completion": self._logs["completion"],
                 **self._logs["rewards"],
+                **self._logs["extra"],
```
[the remark this answers is not in the document]
Comment (manueldeprada) [reply trl/trainer/grpo_trainer.py:2462]: same as previous answer.

**Comment by manueldeprada** [comment]
> Thanks @manueldeprada!
> 
> I understand the need, there's currently no simple way to log extra metrics from within a reward function without modifying the codebase. The closest workaround today is to use separate reward functions with `reward_weights=0` for pure logging, but that doesn't help when the metric is a byproduct of the reward computation and you want to avoid recomputing it.
> 
> That said, I'm not fully sold on the approach. A reward function should ideally be passive: it takes inputs, returns a scalar, and has no side effects. Passing a `log_metric` callback blurs that boundary.
> 
> I don't have a better alternative in mind though, and from the user's perspective, this is probably the most intuitive and simple solution. So I'd consider merging it. Open to thoughts from others. @albertvillanova @AmineDiro.

Thanks for considering @qgallouedec !! 

I agree, its not elegant. But it solves the need with minimal changes. Let me know if you need anything else from my side. On the meantime, it's working well for me!


P.S. On a related note, I’ve been having some issues with Trackio’s local DB–Hub dataset synchronization. For some runs, the data never gets written to the Hub dataset, so I end up losing it. I'm still investigating and not sure yet whether it’s something on my side, with Trackio, or with TRL. I thought I’d mention it here in case anyone has noticed something similar.

**Comment by qgallouedec** [comment]
answering on trackio, yes, it's something that we encountered as well but it should be fixed if you upgrade trackio. we've been working on trackio robustness

**Comment by manueldeprada** [comment]
> answering on trackio, yes, it's something that we encountered as well but it should be fixed if you upgrade trackio. we've been working on trackio robustness

that worked! last trackio requires last huggingface-hub, which requires transformers to 5.0, but that works fine with source TRL as long as I dont use judges.

Updating all these, trackio seems to persists data correctly. Awesome!! 

Another bit you may be interested in: I sometimes get 429 rate-limited for a few minutes while watching https://manueldeprada-thinking.hf.space It's not big issue, but I guess you might want to check that the default simple trackio experience is as smooth as possible.

PS2: cloned TRL works well with sdpa-powered FA4 with latest vllm 😄

**Comment by qgallouedec** [comment]
I don't want to keep this PR opened forever, so I'll go ahead and merge, I think it's actually quite useful.
I'll let the other maintainers suggest reverting if they think of a better pattern. Thanks @manueldeprada!
